src/kudu/consensus/raft_consensus.cc

```diff
@@ -3297,6 +3335,280 @@ void RaftConsensus::SetLeaderDetectedCallback(LeaderDetectedCallback ldcb) {
+  auto prevent_ops_deletion = MakeScopedCleanup([&]() {
+    // Prevent double-deletion of these requests.
+    downstream_request.mutable_ops()->ExtractSubrange(
+      /*start=*/ 0, /*num=*/ downstream_request.ops_size(), /*elements=*/ nullptr);
+  });
+  downstream_request.set_dest_uuid(request->dest_uuid());
+  downstream_request.set_tablet_id(request->tablet_id());
+  downstream_request.set_caller_uuid(request->caller_uuid());
+  downstream_request.set_caller_term(request->caller_term());
+
+  if (request->has_preceding_id()) {
+    *downstream_request.mutable_preceding_id() = request->preceding_id();
+  }
+  if (request->has_committed_index()) {
+    downstream_request.set_committed_index(request->committed_index());
+  }
+  if (request->has_all_replicated_index()) {
+    downstream_request.set_all_replicated_index(request->all_replicated_index());
+  }
+  if (request->has_safe_timestamp()) {
+    downstream_request.set_safe_timestamp(request->safe_timestamp());
+  }
+  if (request->has_last_idx_appended_to_leader()) {
+    downstream_request.set_last_idx_appended_to_leader(request->last_idx_appended_to_leader());
+  }
+
+  downstream_request.set_proxy_caller_uuid(peer_uuid());
+
+  string next_uuid = request->dest_uuid();
+  if (FLAGS_raft_enable_multi_hop_proxy_routing) {
+    RET_RESPOND_ERROR_NOT_OK(routing_table_->NextHop(peer_uuid(), request->dest_uuid(), &next_uuid));
+  }
+
+  if (request->dest_uuid() != next_uuid) {
+    // Multi-hop proxy request.
+    downstream_request.set_proxy_dest_uuid(next_uuid);
+    // Forward the existing PROXY_OP ops.
+    for (int i = 0; i < request->ops_size(); i++) {
+      *downstream_request.add_ops() = request->ops(i);
+    }
+    prevent_ops_deletion.cancel(); // The ops we copy here are not pre-allocated.
+  } else {
+
+    // Reconstitute proxied events from the local cache.
+    // If the cache does not have all events, we retry up until the specified
+    // retry timeout.
+    // TODO(mpercy): Switch this from polling to event-triggered.
+
+    vector<ReplicateRefPtr> messages;
+    do {
+
+      // We assume and enforce that a single request is composed of a range of ops.
+      int64_t first_op_index = -1;
+
+      int64_t max_batch_size = FLAGS_consensus_max_batch_size_bytes - request->ByteSizeLong();
+      for (int i = 0; i < request->ops_size(); i++) {
+        auto& msg = request->ops(i);
+        if (PREDICT_FALSE(msg.op_type() != PROXY_OP)) {
+          RET_RESPOND_ERROR_NOT_OK(Status::InvalidArgument(Substitute(
+              "proxy expected PROXY_OP but received opid {} of type {}",
+              OpIdToString(msg.id()),
+              OperationType_Name(msg.op_type()))));
+        }
+        if (i == 0) {
+          first_op_index = msg.id().index();
+        } else {
+          // TODO(mpercy): It would be nice not to require consecutive indexes in the batch.
+          // We should see if we can support it without a big perf penalty in IOPS.
```

Comment:
I think we need to worry about correctness in how we are picking up messages from the log cache
and there could be race between the message to the proxy node, vs the earlier message to the 
dest uuid in the presence of TRIMs /leader switches. The scenario I am thinking of is when
a leader sends op id 25 and it sends op id 25 as proxy.
Then the leader switches and op id 25 has gotten trimmed or overriden. Is there a chance that the
proxy will use the timer and send the old 25 index first and the trims have not been communicated properly. Maybe you have already thought about these scenarios.

Reply:
No need to worry about fetching the wrong ops. We perform validation on L3540 to make sure the reconstituted ops have the same term and index as the ones the leader specified.